lexicon lang spec clarifications (#463)
bluesky-social/atproto-website: src/app/[locale]/specs/lexicon/en.mdx

Note that unlike `object`, there is no `nullable` field on `params`.

> `params` can not be declared as top-level named types in a schema `defs` array. This means that they can not be reference by a `ref` or `union`.

Reviewer: ```suggestion
`params` can not be declared as top-level named types in a schema `defs` array. This means that they can not be referenced by a `ref` or `union`.
```

> `params` can not be declared as top-level named types in a schema `defs` array. This means that they can not be reference by a `ref` or `union`.

Reviewer: also i know `params` is technically plural, but this reads weird to me and when referred to as a type, I think we should treat it has singular. for instance: "`params`  can not be declared as a top-level named type..."

Tokens are similar to the concept of a "symbol" in some programming languages, distinct from strings, variables, built-in keywords, or other identifiers. For example, tokens could be defined to represent the state of an entity (in a state machine), or to enumerate a list of categories.

### `ref`

Type-specific fields:

- `ref` (string, required): reference to another schema definition

Refs are a mechanism for re-using a schema definition in multiple places. The intended use is to reduce duplication of definitions for types like `object` or a `string` with specific restrictions.

The `ref` string can be a global reference to a Lexicon type definition (an NSID, optionally with a `#`-delimited name indicating a definition other than `main`), or can indicate a local definition within the same Lexicon file (a `#` followed by a name).

When a `ref` points to an `object` type, the type of the encoded data is unambiguous. This means that the `$type` field should not be included in encoded data as a discriminator. This does not apply to references to `record` types, which always include the `$type` discriminator.

A `ref` can not point to a `token` type.

> Refs can not be declared as top-level named types in a schema `defs` array. This means that a `ref` can not point to another `ref`, nor a `union`.

Reviewer 2: ```suggestion
Refs can not be declared as top-level named types in a schema `defs` map. This means that a `ref` can not point to another `ref`, nor a `union`. They can reference container types that contain `ref` types.
```

By default unions are "open", meaning that future revisions of the schema could add more types to the list of refs (though can not remove types). This means that implementations should be permissive when validating, in case they do not have the most recent version of the Lexicon. The `closed` flag (boolean) can indicate that the set of types is fixed and can not be extended in the future.

A `union` schema definition with no `refs` is allowed and similar to `unknown`, as long as the `closed` flag is false (the default). The main difference is that the data is required to include the `$type` field.

A closed union with only a single type referenced is similar to `ref`. The main difference is that the data is required to include the `$type` field.

An empty closed union (a union with an empty refs list and the `closed` flag set to true) is an invalid schema.

> The schema definitions pointed to by a `union` must have type `object` or types with a clear mapping to an object, like a `record`. All the variants must be represented by a CBOR map (or JSON Object) and must include a `$type` field indicating the variant type. Because the data must be an object, unions can not reference `token` (which would correspond to string data).

Reviewer: I would rejigger both of these paragraphs:

The schema definitions referenced to by a `union` must have type `object` or `record`. All the variants must be represented by a CBOR map (or JSON Object) and must include a `$type` field indicating the variant type.  A `union` can not point to types `query`, `procedure`, `subscription`, `unknown`, `blob`, `cid-link`, `array`, `params`, `ref`, `union`, `token` or any other currently defined types.

> To be more complete, `union` references must point to a type `object` or `record`. They can not point to types `query`, `procedure`, `subscription`, `unknown`, `blob`, `cid-link`, `array`, `params`, `ref`, `union`, or any other currently defined types.

Reviewer: maybe worth including `token` in this list as well

Author: they were mentioned in the paragraph just above, but agree it is best to have them in the same list.

> Unions can not be declared as top-level named types in a schema `defs` array. This means that a `union` can not reference another `union`, nor a `ref`.

Reviewer 2: ```suggestion
Unions can not be declared as top-level named types in a schema `defs` map. This means that a `union` can not reference another `union`, nor a `ref`.
```
